src/GcsRestore.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Keboola\ProjectRestore;
+
+use Exception;
+use Keboola\StorageApi\Client as StorageApi;
+use Keboola\Temp\Temp;
+use Psr\Log\LoggerInterface;
+
+class GcsRestore extends Restore
+{
+    public function __construct(
+        StorageApi $sapiClient,
+        readonly array $listFiles,
+        ?LoggerInterface $logger = null,
+    ) {
+        parent::__construct($sapiClient, $logger);
+    }
+
+    /**
+     * @return resource|string
+     */
+    protected function getDataFromStorage(string $filePath, bool $useString = true)
+    {
+        $tmp = new Temp();
+
+        $targetFile = $tmp->createFile(uniqid());
+
+        $this->copyFileFromStorage($filePath, $targetFile->getPathname());
+
+        if (!$useString) {
+            $file = fopen($targetFile->getPathname(), 'r');
+            if (!$file) {
+                throw new Exception(sprintf(
+                    'File "%s" does not exists.',
+                    $filePath,
+                ));
+            }
+            return $file;
+        } else {
+            return (string) file_get_contents($targetFile->getPathname());
+        }
+    }
+
+    protected function copyFileFromStorage(string $sourceFilePath, string $targetFilePath): void
+    {
+        $explodedSourceFilePath = explode('/', $sourceFilePath);
+        $sourceUrl = $this->findUrl($this->listFiles, $explodedSourceFilePath);
+        copy($sourceUrl, $targetFilePath);
+    }
+
+    protected function listComponentConfigurationsFiles(string $filePath): array
+    {
+        $explodedPath = explode('/', $filePath);
+        array_shift($explodedPath);
+
+        $actualList = $this->listFiles['configurations'];
+        return $this->findPartOfTree($actualList, $explodedPath);
+    }
+
+    protected function listTableFiles(string $tableId): array
+    {
+        $explodedTableId = explode('.', $tableId);
+        $tablePath = $explodedTableId;
+        array_pop($tablePath);
+
+        $actualList = $this->findPartOfTree($this->listFiles, $explodedTableId);
+
+        $table = end($explodedTableId);
+
+        $parts = [];
+        foreach ($actualList as $fileName => $url) {
+            if (str_starts_with($fileName, $table . '.')) {
+                $parts[] = implode('/', $tablePath) . '/' . $fileName;
+            }
+        }
+        return $parts;
+    }
+
+    public function findPartOfTree(array $list, array $findPath): array
+    {
+        return (array) $this->findInArray($list, $findPath);
+    }
+
+    private function findUrl(array $list, array $findPath): string
+    {
+        $result = $this->findInArray($list, $findPath);
+
+        if (is_string($result)) {
+            return $result;
+        }
+
+        throw new Exception(sprintf(
+            'File "%s" does not exists.',
+            implode('/', $findPath),
+        ));
+    }
+
+    private function findInArray(array $list, array $findPath): string|array
```

Comment:
Tohle je prapodivný 😄 Nešlo by to chování pro příští generace zdokumentovat?

Reply:
máš ho tam :)

Reply:
Jen teda jestli bude nalezená položka `int` jako v tom doc, tak to skape na return type `string|array` 🙂 

A když na to tak čumím, nešlo by to zjednodušit?

```php
private function findInArray(array $array, array $path): mixed
{
    foreach ($path as $key) {
        if (!isset($array[$key])) {
            break;
        }
        $array = $array[$key];
    }
    return $array;
}
```

Reply:
no nikdy to nebude int... je to seznam URL adres, ale jako máš pravdu :D 

a jo, to by šlo 👍 pošlu to tam

Reply:
https://github.com/keboola/php-kbc-project-restore/pull/56/commits/7410dbee851b61dac316e3fbcda1c1f0c8b81fef